rust/chains/hyperlane-ethereum/src/ism.rs

```diff
@@ -0,0 +1,115 @@
+#![allow(clippy::enum_variant_names)]
+#![allow(missing_docs)]
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use ethers::providers::Middleware;
+use tracing::instrument;
+
+use hyperlane_core::{
+    ChainResult, ContractLocator, HyperlaneAbi, HyperlaneChain, HyperlaneContract, HyperlaneDomain,
+    HyperlaneProvider, Ism, H256,
+};
+
+use crate::contracts::i_interchain_security_module::{
+    IInterchainSecurityModule as EthereumIsmInternal, IINTERCHAINSECURITYMODULE_ABI,
+};
+use crate::trait_builder::BuildableWithProvider;
+use crate::EthereumProvider;
+
+impl<M> std::fmt::Display for EthereumIsmInternal<M>
+where
+    M: Middleware,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+pub struct IsmBuilder {}
+
+#[async_trait]
+impl BuildableWithProvider for IsmBuilder {
+    type Output = Box<dyn Ism>;
+
+    async fn build_with_provider<M: Middleware + 'static>(
+        &self,
+        provider: M,
+        locator: &ContractLocator,
+    ) -> Self::Output {
+        Box::new(EthereumIsm::new(Arc::new(provider), locator))
+    }
+}
+
+/// A reference to an Ism contract on some Ethereum chain
+#[derive(Debug)]
+pub struct EthereumIsm<M>
+where
+    M: Middleware,
+{
+    contract: Arc<EthereumIsmInternal<M>>,
+    domain: HyperlaneDomain,
+}
+
+impl<M> EthereumIsm<M>
+where
+    M: Middleware + 'static,
+{
+    /// Create a reference to a mailbox at a specific Ethereum address on some
+    /// chain
+    pub fn new(provider: Arc<M>, locator: &ContractLocator) -> Self {
```

Comment:
If you make this new fn `async` you can set the ISM type at init.

Reply:
Why is this better than calling it in metadata builder? IIUC if we put it here we need to duplicate the u8->enum logic for each implementation of the trait

Reply:
I guess the benefit is that if you query it in the constructor you can enforce that the async call is only made once.

Reply:
Yes, that

Reply:
The way things are currently structured there's no reason it would ever be called twice, so I'm not sure this is necessary. IMO there's something nice about all the constructors acting the same way (which does not involve contract calls).